### ISSUE TITLE
Support parallel Maven builds

### ISSUE THREAD

**Issue by JanecekPetr** [comoyo/commons#36]
First, let me say THANK YOU for such a good plugin. Finally someone made a working protoc maven plugin that behaves as it should.

One minor quirk, though: When running my build in parallel (https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3), Maven complains that:

    [INFO] ------------------------------------------------------------------------
    [INFO] Building XXX 00.00.1-SNAPSHOT
    [INFO] ------------------------------------------------------------------------
    [WARNING] *****************************************************************
    [WARNING] * Your build is requesting parallel execution, but project      *
    [WARNING] * contains the following plugin(s) that are not marked as       *
    [WARNING] * @threadSafe to support parallel building.                     *
    [WARNING] * While this /may/ work fine, please look for plugin updates    *
    [WARNING] * and/or request plugins be made thread-safe.                   *
    [WARNING] * If reporting an issue, report it against the plugin in        *
    [WARNING] * question, not against maven-core                              *
    [WARNING] *****************************************************************
    [WARNING] The following plugins are not marked @threadSafe in XXX:
    [WARNING] com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47
    [WARNING] *****************************************************************

Could it be made sure that this plugin works well in parallel and if yes, be marked as such, please?

**Comment by argggh** [comment]
Thank you for the kind words!  I would assume that this plugin should be threadsafe, but I'd need to take a proper look.  I'll go see.

**Comment by JanecekPetr** [comment]
Hello!
Any update on this? If the plugin actually _is_ thread-safe, then marking it as such should be a quick job. I don't want you to rewrite it if it's currently unsafe, of course, I know people tend to be busy doing some actual work...

**Comment by argggh** [comment]
My apologies, I started looking into this, and as far as I can tell there's no reason the plugin itself should not be threadsafe.  However, it does depend on `org.apache.maven.repository.RepositorySystem`'s `resolve` also being threadsafe, and at that point real work intervened...  I'll try to dig some more into this.

**Comment by JanecekPetr** [comment]
Hi again!

I've been testing this for the past few days as it seemed to work correctly. However, I've seen a failure or two. Basically the results are exactly as you say:

1. Maven 3.2.3: both locally (Windiws 7) and on our build servers (RHEL 6.5), this happens upon the first build:

        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Using protoc com.comoyo.maven.plugins:protoc-bundled-plugin:exe:2.4.1-linux-x86_64:1.4.47
        [INFO] Downloading: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe
        [INFO] Downloaded: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe (2497 KB at 2438.2 KB/sec)
        [INFO] Compiling /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf to /xxxxxx/xxxxxx/xxxxxx/target/generated-sources/protobufain]
        [INFO] Downloaded: http://xxxxxx:8080/nexus/content/groups/main/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe (2497 KB at 2837.2 KB/sec)
        Failed to execute goal com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47:run (default) on project xxxxxx: Unable to compile /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf/xxxxxx.proto

        [ERROR] Failed to execute goal com.comoyo.maven.plugins:protoc-bundled-plugin:1.4.47:run (default) on project xxxxxx: Unable to compile /xxxxxx/xxxxxx/xxxxxx/src/main/protobuf/xxxxxx.proto: Cannot run program "/xxxxxx/xxxxxx/m2-repository/com/comoyo/maven/plugins/protoc-bundled-plugin/1.4.47/protoc-bundled-plugin-1.4.47-2.4.1-linux-x86_64.exe": error=13, Permission denied -> [Help 1]

2. When using Maven 3.2.5 or newer (I tried both 3.3.3 and 3.3.9), I could not make it fail. Simultaneous downloads of protoc seem to work fine, only one thread actually downloads and the others simply wait and then report a successful download of 0 B at 0 KB/s.


Therefore, this is clearly a Maven bug and you don't need to care about it too much (a README notice would be awesome, though). The rest works fine, and therefore this plugin deserves the `@threadSafe` annotation. And 2.6.1 protoc.

Thank you!

**Comment by argggh** [comment]
Ah, nice to have that confirmed.  I guess it is arguable if it is a Maven problem or a problem with the sequence `resolve` ... `setExecutable` being inherently race vulnerable.  It should probably be sufficient to slap a `synchronized` declaration on `resolveProtocArtifact`, so I think I'll just do that and then mark the plugin as threadsafe. 

To be honest, I thought 2.6.1 was in 1.4.50 of this plugin, but obviously I'm wrong about that.  It's on master, but not released. :(  I'll get to that as well.

**Comment by argggh** [comment]
Have a look at 1.4.53, I think these issues should be addressed there.

**Comment by JanecekPetr** [comment]
Nope. Still the same issue. I guess (and it's just a guess, I haven't ever written any maven plugin) that Maven creates multiple Mojo instances (with different config) and runs them on different modules in parallel. That means that you actually need to lock on something static instead of the Mojo instance...

**Comment by argggh** [comment]
That is a shame (and a surprise, given the instructions I find on how to make your maven plugins threadsafe).  I will try to investigate some more, but I think I need to reproduce the issue you are seeing locally before I can make headway on that.  Is the project in question open to the public?

**Comment by JanecekPetr** [comment]
Looking at: [Mojo thread safety assertion checklist](https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3#ParallelbuildsinMaven3-Mojothreadsafetyassertionchecklist), you do indeed need a static lock as there are multiple Mojo instances at the same time.

Unfortunately our project is private. The repro is very easy though, I might put something together in the evening. You want to trigger simultaneous protoc downloads. Works like this:

- parent POM
    - 1. child project with some protobufs to compile
    - 2. child project with some protobufs to compile
    - 3. child project that depends on both 1. and 2.

And then, with Maven < 3.2.5:

    cd parent
    mvn clean package -T 1C

Boom.

**Comment by argggh** [comment]
Ah, I missed that nuance.  Still, as long as there's only JVM that shouldn't be a problem, we just need to slightly adjust the code.  Any example repos that you are able to contribute would be helpful, otherwise I'll look at it myself.

**Comment by JanecekPetr** [comment]
https://github.com/JanecekPetr/boom-proto
Download, delete the compiler from your local repo (!)

    mvn clean generate-sources -T 1C -V

Crashes half of the times on my PC.
